storagenode/pieces/cache.go

```diff
@@ -0,0 +1,163 @@
+	store.cacheMutex.Lock()
+	defer store.cacheMutex.Unlock()
+	return store.cache.totalBySatellite[satelliteID]
+}
+
+// UpdateCache updates the live used space totals
+// with a pieceSize that was either created or deleted where the pieceSize is
+// only the content size and does not include header bytes
+func (store *StoreWithCache) UpdateCache(ctx context.Context, satelliteID storj.NodeID, pieceSize int64) {
+	store.cacheMutex.Lock()
+	defer store.cacheMutex.Unlock()
+	store.cache.total += pieceSize
+	store.cache.totalBySatellite[satelliteID] += pieceSize
+}
+
+// RecalculateSpaceUsedCache sums up the bytes for all pieces (not including headers) currently stored.
+// The live values for used space are initially calculated when the storagenode starts up, then
+// incrememted/decremeted when pieces are created/deleted. In addition we want to occasionally recalculate
+// the live values to confirm correctness. This method RecalculateSpaceUsedCache is responsible for doing that.
+func (store *StoreWithCache) RecalculateSpaceUsedCache(ctx context.Context) error {
+	store.cacheMutex.Lock()
+	// Save the current live values before we start recalculating
+	// so we can compare them to what we recalculate
+	spaceUsedWhenIterationStarted := store.cache
+	store.cacheMutex.Unlock()
+
+	spaceUsedResultOfIteration, err := store.spaceUsedForPiecesAndBySatellitesSlow()
+	if err != nil {
+		return err
+	}
+
+	store.cacheMutex.Lock()
+	defer store.cacheMutex.Unlock()
+	// Since it might have taken a long time to iterate over all the pieces, here we need to check if
+	// we missed any writes/deletes of pieces while we were iterating and estimate
+	store.estimateAndSave(spaceUsedWhenIterationStarted, spaceUsedResultOfIteration)
+
+	return nil
+}
+
+func (store *StoreWithCache) estimateAndSave(oldSpaceUsed, newSpaceUsed spaceUsedCache) {
+	estimatedTotalsBySatellites := map[storj.NodeID]int64{}
+	for satelliteID, newSpaceUsedTotal := range newSpaceUsed.totalBySatellite {
+		estimatedTotalsBySatellites[satelliteID] = estimateRecalculation(newSpaceUsedTotal,
+			oldSpaceUsed.totalBySatellite[satelliteID],
+			store.cache.totalBySatellite[satelliteID],
+		)
+	}
+
+	store.cache.totalBySatellite = estimatedTotalsBySatellites
+	store.cache.total = estimateRecalculation(newSpaceUsed.total,
+		oldSpaceUsed.total,
+		store.cache.total,
+	)
+}
+
+func estimateRecalculation(newSpaceUsedTotal, spaceUsedWhenIterationStarted, spaceUsedWhenIterationEnded int64) int64 {
+	if newSpaceUsedTotal == spaceUsedWhenIterationEnded {
+		return newSpaceUsedTotal
+	}
+
+	// If we missed writes/deletes while iterating, we will assume that half of those missed occurred before
+	// the iteration and half occurred after. So here we add half of the delta to the result space used totals
+	// from the iteration to account for those missed.
+	spaceUsedDeltaDuringIteration := spaceUsedWhenIterationStarted - spaceUsedWhenIterationEnded
```

Comment:
It's late and I'm totally not sure, but is this backward? I think we want `spaceUsedWhenIterationEnded - spaceUsedWhenIterationStarted`, because a positive delta should mean an increase in space used (since we add half of it to `newSpaceUsedTotal`)

Reply:
hmm good catch. let me double check which way it should be